### PR TITLE
EZP-26885: Refactor Field Types external storages to inject gateways and connection handler via DI

### PR DESCRIPTION
**Pull request by alongosz**
Part 1 of [EZP-26885](https://jira.ez.no/browse/EZP-26885) implementation.
------------

> We want to be able to use `\Doctrine\DBAL\Connection` for FieldTypes external storage directly.
> To achieve this, FieldType external storage needs to be injected with `DoctrineStorage` gateway which uses Doctrine instead of Legacy Connection Handler.

This PR is the first step towards solving [EZP-26885](https://jira.ez.no/browse/EZP-26885). To be able easily inject Doctrine instead of Legacy DatabaseHandler we need to make the latter fully injectable. Up to now, external storage gateway was chosen by context passed by `\eZ\Publish\Core\Persistence\Legacy\Content\StorageHandler`.

This PR refactors Field Type-specific external storages so they get injected with proper, FT-specific gateways. 
Each gateway gets injected with `\eZ\Publish\Core\Persistence\Doctrine\ConnectionHandler`.
In the future Legacy gateway injected into FT-specific external storage can be replaced with Doctrine gateway which would use `\Doctrine\DBAL\Connection`.

**TODO**:
- [x] Introduce to SPI layer `GatewayBasedStorage` which refactored FT storages should extend.
- [x] Introduce to SPI layer `StorageGateway` to be extended by refactored FT ext. storages gateways.
- [x] Deprecate `\eZ\Publish\Core\FieldType\GatewayBasedStorage` which sets and gets gateways from context.
- [x] Deprecate `\eZ\Publish\Core\FieldType\StorageGateway` which allows setting connection manually, without DI.
- [x] Refactor `Keyword` FT external storage.
- [x] Refactor `BinaryFile` and `Media` FT external storage.
- [x] Refactor `Image` FT external storage.
- [x] Refactor `MapLocation` FT external storage.
- [x] Refactor `User` FT external storage.
- [x] Refactor `Page` FT external storage.
- [x] Refactor `RichText` FT external storage.
- [x] Refactor `Url` FT external storage.
- [x] Perform manual tests with web app.
- [x] Make upgrade doc.

### CODE REVIEW

**Comment by alongosz** [comment]
Example of actual implementation of DoctrineStorage gateway is available in the PR #1993.
For Keyword FT: 46d5589.

**Comment by alongosz** [comment]
review ping @andrerom @kore @Nattfarinn

**Comment by kore** [comment]
+1

File: doc/upgrade/6.10.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
+# Upgrade steps from 6.9 to 6.10
```
Comment (andrerom) [inline doc/upgrade/6.10.md]: 6.11 by now

File: doc/upgrade/6.10.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+## Field Type external storage
+
+### External storage gateways
+
+Abstract base class `eZ\Publish\Core\FieldType\StorageGateway` for the gateway of a Field Type external storage and the
+`eZ\Publish\Core\FieldType\StorageGateway::setConnection` method are deprecated.
+
+Extend the `eZ\Publish\SPI\FieldType\StorageGateway` class and inject Database Handler directly
```
Comment (andrerom) [inline doc/upgrade/6.10.md]: to better signal that the change is not something field types have to do immediately, start the sentence by saying something like:

`To prepare your FieldType for future storage engines and avoid using the deprecated abstract, extend the ..`

File: doc/upgrade/6.10.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+    // ...
+}
+```
+### External storage handlers
+
+Abstract base class `eZ\Publish\Core\FieldType\GatewayBasedStorage` for Field Type external storage
+which uses gateway is deprecated.
+
+Use `eZ\Publish\SPI\FieldType\GatewayBasedStorage` instead and inject external
```
Comment (andrerom) [inline doc/upgrade/6.10.md]: similar comment here

File: eZ/Publish/Core/FieldType/GatewayBasedStorage.php [hunk 4]
```diff
@@ -39,6 +41,16 @@
      */
     public function __construct(array $gateways = array())
     {
+        @trigger_error(
+            sprintf(
+                '%s extends deprecated %s. Extend %s instead',
+                static::class,
+                self::class,
+                \eZ\Publish\SPI\FieldType\GatewayBasedStorage::class
```
Comment (andrerom) [inline eZ/Publish/Core/FieldType/GatewayBasedStorage.php]: place in use section to not use FQCN inline

File: eZ/Publish/Core/FieldType/GatewayBasedStorage.php [hunk 5]
```diff
@@ -15,6 +15,8 @@
 /**
  * Storage gateway base class to be used by FieldType storages.
  *
+ * @deprecated Since 6.10. Use {@link \eZ\Publish\SPI\FieldType\GatewayBasedStorage}
```
Comment (andrerom) [inline eZ/Publish/Core/FieldType/GatewayBasedStorage.php]: 6.11, same everywhere else in this PR

**Comment by andrerom** [comment]
Another thing, this deprecation should also be mentioned with a point in doc/bc/6.11 with reference to the upgrade doc for further info.

File: doc/bc/changes-6.11.md [hunk 6]
```diff
@@ -0,0 +1,20 @@
+# Backwards compatibility changes
+
+Changes affecting version compatibility with former or future versions.
+
+## Changes
+
+- EZP-26885: Field Type external storage
+  - Field Type external storage gets directly injected with a proper external storage gateway.
```
Comment (andrerom) [inline doc/bc/changes-6.11.md]: This is misleading isn't it, people will need to migrate as stated in upgrade doc in order for this change to affect them. As in if they stay on old abstract nothing will change or?

Comment (alongosz) [inline doc/bc/changes-6.11.md]: Ok, I wasn't sure how to indicate this. Maybe I just should keep `## Changes` section empty and keep `## Deprecations` only?

Comment (andrerom) [inline doc/bc/changes-6.11.md]: yes I think so

we almost always first deprecate, and in a later release change or remove ;)

Comment (alongosz) [inline doc/bc/changes-6.11.md]: Done, all docs in 13a9dc2

File: doc/bc/changes-6.11.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
+# Backwards compatibility changes
+
+Changes affecting version compatibility with former or future versions.
+
+## Changes
+
+## Deprecations
+
+- EZP-26885: Field Type external storage
+  - Abstract base class `eZ\Publish\Core\FieldType\StorageGateway` for the gateway of a Field Type
```
Comment (andrerom) [inline doc/bc/changes-6.11.md]: Should you also list up `eZ\Publish\Core\FieldType\GatewayBasedStorage` here?

Comment (alongosz) [inline doc/bc/changes-6.11.md]: Yeah, sorry :( Somehow I thought we have only two deprecations in Core. I looked over entire diff and in fact we have 4 in total. Added as a fixup in 23722a4 for easier review.

Comment (alongosz) [inline doc/bc/changes-6.11.md]: Amended in a6d83a7

**Comment by andrerom** [comment]
Great work @alongosz, merging :)

You can now rebase the other one when you have the time ;)

**Pull request merged**